# Question 3: KNN + Random Forest Classification

## Part A: Classification : KNN & Random Forest

In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.neighbors import KNeighborsClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, precision_score, recall_score

# Load the dataset
df = pd.read_csv(r'C:\Users\ihsan\Downloads\priceClassification\train.csv')

# Separate features and target
X = df.drop('price_range', axis=1)
y = df['price_range']

# Split the dataset into training and testing sets (80% train, 20% test)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Scale the features
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)


### 1. KNN (K=3) and KNN (K=57)

In [2]:
# KNN with K=3
knn_3 = KNeighborsClassifier(n_neighbors=3)
knn_3.fit(X_train_scaled, y_train)
y_pred_3 = knn_3.predict(X_test_scaled)

acc_3 = accuracy_score(y_test, y_pred_3)
prec_3 = precision_score(y_test, y_pred_3, average='weighted')
rec_3 = recall_score(y_test, y_pred_3, average='weighted')

print(f"KNN (K=3) - Accuracy: {acc_3:.4f}, Precision: {prec_3:.4f}, Recall: {rec_3:.4f}")


KNN (K=3) - Accuracy: 0.4825, Precision: 0.5183, Recall: 0.4825


In [3]:
# KNN with K=57
knn_57 = KNeighborsClassifier(n_neighbors=57)
knn_57.fit(X_train_scaled, y_train)
y_pred_57 = knn_57.predict(X_test_scaled)

acc_57 = accuracy_score(y_test, y_pred_57)
prec_57 = precision_score(y_test, y_pred_57, average='weighted')
rec_57 = recall_score(y_test, y_pred_57, average='weighted')

print(f"KNN (K=57) - Accuracy: {acc_57:.4f}, Precision: {prec_57:.4f}, Recall: {rec_57:.4f}")


KNN (K=57) - Accuracy: 0.6125, Precision: 0.6563, Recall: 0.6125


**Which value of K gives the best result? Justify your choice in 1–2 lines.**

*Answer:* K=57 gives a better result because it has higher accuracy (0.6125) compared to K=3 (0.4825). A larger K helps smooth out noise in the training data, leading to a better-generalizing decision boundary.

### 2. Random Forest Classifier

In [4]:
# Random Forest Classifier
rf = RandomForestClassifier(random_state=42)
rf.fit(X_train_scaled, y_train)
y_pred_rf = rf.predict(X_test_scaled)

acc_rf = accuracy_score(y_test, y_pred_rf)
prec_rf = precision_score(y_test, y_pred_rf, average='weighted')
rec_rf = recall_score(y_test, y_pred_rf, average='weighted')

print(f"Random Forest - Accuracy: {acc_rf:.4f}, Precision: {prec_rf:.4f}, Recall: {rec_rf:.4f}")


Random Forest - Accuracy: 0.8925, Precision: 0.8961, Recall: 0.8925


### 3. Comparison Table

| Model | Accuracy | Precision | Recall |
| :--- | :--- | :--- | :--- |
| **KNN (K=3)** | 0.4825 | 0.5183 | 0.4825 |
| **KNN (K=57)** | 0.6125 | 0.6563 | 0.6125 |
| **Random Forest** | 0.8925 | 0.8961 | 0.8925 |

**Which model performs best overall?**

*Answer:* The Random Forest model performs the best overall as it achieves the highest accuracy, precision, and recall.

## Part B: Conceptual 

### 1. Precision and Recall Analysis

**What does Recall tell you about a model's performance?**
*Answer:* Recall (or Sensitivity) measures the proportion of actual positive cases that the model correctly identified. It indicates how effectively the model finds all the positive instances without missing them.

**What does Precision tell you about a model's performance?**
*Answer:* Precision measures the proportion of positive predictions that were actually correct. It indicates how trustworthy the model is when it predicts a positive case, highlighting the rate of false alarms.

**If you had to choose one model for detecting a disease (where missing a positive case is very costly), which would you choose and why?**
*Answer:* We would choose the model with higher **Recall**. Missing a positive case is costly, so it's critical to minimize false negatives and catch every possible positive case, even at the expense of having more false alarms (lower precision).

### 2. Random Forest vs Decision Tree

**Why does Random Forest generally perform better than a single Decision Tree? Explain in 2–3 lines.**
*Answer:* A single Decision Tree is prone to overfitting to the training data. A Random Forest combines multiple independent decision trees trained on random subsets of the data and features, which reduces variance and overfitting, leading to much better generalization on unseen data.

